bayes_kit/drghmc.py

```diff
@@ -0,0 +1,437 @@
+    current sample using the leapfrog integrator. The probability of accepting this
+    transition is recursively computed. If accepted, DrGhmcDiag returns the proposed
+    sample and its log density. If rejected, DrGhmcDiag proposes a new sample generated
+    with new stepsize and number of steps for the leapfrog integrator. This process is
+    repeated until a maximum number of proposals is reached or probabilistic delayed
+    rejection terminates early.
+
+    Intuitively, DrGhmcDiag reduces the stepsize in subsequent proposals if the
+    previous proposal was rejected for better sampling resolution of high-curvature
+    state space regions.
+
+    Implementation based on Modi, Chirag, Alex Barnett, and Bob Carpenter. "Delayed
+    rejection Hamiltonian Monte Carlo for sampling multiscale distributions." Bayesian
+    Analysis 1.1 (2023): 1-28. While the paper represents position and momentum with
+    (p, q), this implementation uses (theta, rho) for consistency with bayes-kit. All
+    computation is done on the log scale unless explicitly stated otherwise.
+    """
+
+    def __init__(
+        self,
+        model: GradModel,
+        stepsize: Union[int | float, list[int | float], Callable[[int], int | float]],
+        steps: Union[int, list[int], Callable[[int], int]],
+        metric_diag: Optional[VectorType] = None,
+        init: Optional[VectorType] = None,
+        seed: Optional[Seed] = None,
+        num_proposals: int = 3,
+        probabilistic: bool = True,
+        dampening: float = 1.0,
+    ):
+        """Initialize the DrGhmcDiag sampler.
+
+        Args:
+            model: model with log density and gradient
+            stepsize: stepsize in each leapfrog step. Defaults to None.
+            steps: number of leapfrog steps. Defaults to None.
+            metric_diag: diagonal of the metric. Defaults to None.
+            init: initialize Markov chain. Defaults to None.
+            seed: seed for numpy rng. Defaults to None.
+            num_proposals: number of delayed rejection proposals. Defaults to 3.
+            probabilistic: whether to use probabilistic delayed rejection. Defaults to True.
+            dampening: GHMC momentum dampening factor in [0, 1] inclusive. Defaults to 1.0.
+        """
+        self._model = model
+        self._dim = self._model.dims()
+        self._num_proposals = num_proposals
+        self._stepsize_list = self._init_stepsize(stepsize)
+        self._steps_list = self._init_steps(steps)
+        self._metric = metric_diag or np.ones(self._dim)
+        self._rng = np.random.default_rng(seed)
+        self._theta = (
+            init
+            if (init is not None and init.shape != (0,))
+            else self._rng.normal(size=self._dim)
+        )
+        self._rho = self._rng.normal(size=self._dim)
+        self._probabilistic = probabilistic
+        if not 0 <= dampening <= 1:
+            raise ValueError(f"dampening must be within [0, 1] inclusive")
```

Comment:
True, but it doesn't cost anything to state it explicitly, and can make it more accessible.

Reply:
The domain should actually be $(0,1]$ -- I made a mistake.

I am removing the word inclusive because both endpoints of the domain are now *not* inclusive.